man/ipmipower.8
.\"#############################################################################
.\"$Id: ipmipower.8.pre.in,v 1.59 2010-06-30 21:56:36 chu11 Exp $
.\"#############################################################################
.\"  Copyright (C) 2007-2012 Lawrence Livermore National Security, LLC.
.\"  Copyright (C) 2003-2007 The Regents of the University of California.
.\"  Produced at Lawrence Livermore National Laboratory (cf, DISCLAIMER).
.\"  Written by Albert Chu <chu11@llnl.gov>
.\"  UCRL-CODE-155698
.\"  
.\"  This file is part of Ipmipower, a remote power control utility. 
.\"  For details, see http://www.llnl.gov/linux/.
.\"
.\"  Ipmipower is free software; you can redistribute it and/or modify it under
.\"  the terms of the GNU General Public License as published by the Free
.\"  Software Foundation; either version 3 of the License, or (at your option)
.\"  any later version.
.\"  
.\"  Ipmipower is distributed in the hope that it will be useful, but WITHOUT 
.\"  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
.\"  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License 
.\"  for more details.
.\"  
.\"  You should have received a copy of the GNU General Public License along
.\"  with Ipmipower.  If not, see <http://www.gnu.org/licenses/>.
.\"############################################################################
.TH ipmipower 8 "2012-05-17" "ipmipower 1.1.5" "System Commands"
.SH "NAME"
ipmipower \- IPMI power control utility
.SH "SYNOPSIS"
.B ipmipower
[\fIOPTION\fR...]
.SH "DESCRIPTION"
.B ipmipower
allows users to remotely power on, off, cycle, hard reset, get a power
status query, perform a pulse diagnostic interrupt, or initiate a
soft-shutdown of the OS via ACPI through the IPMI over LAN protocol.
.LP
When a power command (--on, --off, --cycle, --reset, --stat, --pulse,
or --soft) is specified on the command line,
.B ipmipower
will attempt to run the power command on all hostnames listed on the
command line then exit.
.LP
If no power commands are specified on the command line,
.B ipmipower
will run in interactive mode. Interactive mode gives the user a
command line interface to enter various commands. Details of the
interactive command line interface can be found below under
INTERACTIVE COMMANDS.
.LP
.LP
Listed below are general IPMI options, tool specific options, trouble
shooting information, workaround information, examples, and known
issues. For a general introduction to FreeIPMI please see
.B freeipmi(7).
.SH "GENERAL OPTIONS"
The following options are general options for configuring IPMI
communication and executing general tool commands.
.TP
\fB\-D\fR \fIIPMIDRIVER\fR, \fB\-\-driver\-type\fR=\fIIPMIDRIVER\fR
Specify the driver type to use instead of doing an auto selection.
The currently available outofband drivers are LAN and LAN_2_0, which
perform IPMI 1.5 and IPMI 2.0 respectively.
.TP
\fB\-h\fR \fIIPMIHOST1,IPMIHOST2,...\fR, \fB\-\-hostname\fR=\fIIPMIHOST1,IPMIHOST2,...\fR
Specify the remote host(s) to communicate with. Multiple hostnames
may be separated by comma or may be specified in a range format; see
HOSTRANGED SUPPORT below.
.TP
\fB\-u\fR \fIUSERNAME\fR, \fB\-\-username\fR=\fIUSERNAME\fR
Specify the username to use when authenticating with the remote host.
If not specified, a null (i.e. anonymous) username is assumed. The
user must have atleast OPERATOR privileges to run the --on, --off,
--reset, --cycle, --pulse, or --soft power control commands. The user
must have atleast USER privileges to determine the power status of the
machine through --stat.
.TP
\fB\-p\fR \fIPASSWORD\fR, \fB\-\-password\fR=\fIPASSWORD\fR
Specify the password to use when authenticationg with the remote host.
If not specified, a null password is assumed. Maximum password length
is 16 for IPMI 1.5 and 20 for IPMI 2.0.
.TP
\fB\-P\fR, \fB\-\-password-prompt\fR
Prompt for password to avoid possibility of listing
it in process lists.
.TP
\fB\-k\fR \fIK_G\fR, \fB\-\-k-g\fR=\fIK_G\fR
Specify the K_g BMC key to use when authenticating with the remote
host for IPMI 2.0. If not specified, a null key is assumed. To input
the key in hexadecimal form, prefix the string with '0x'. E.g., the
key 'abc' can be entered with the either the string 'abc' or the
string '0x616263'
.TP
\fB\-K\fR, \fB\-\-k-g-prompt\fR
Prompt for k-g to avoid possibility of listing it in process lists.
.TP
\fB\-\-session-timeout\fR=\fIMILLISECONDS\fR
Specify the session timeout in milliseconds. Defaults to 20000
milliseconds (20 seconds) if not specified.
.TP
\fB\-\-retransmission-timeout\fR=\fIMILLISECONDS\fR
Specify the packet retransmission timeout in milliseconds. Defaults
to 400 milliseconds (0.4 seconds) if not specified.
.TP
\fB\-a\fR \fIAUTHENTICATION\-TYPE\fR, \fB\-\-authentication\-type\fR=\fIAUTHENTICATION\-TYPE\fR
Specify the IPMI 1.5 authentication type to use. The currently
available authentication types are NONE, STRAIGHT_PASSWORD_KEY, MD2,
and MD5. Defaults to MD5 if not specified.
.TP
\fB\-I\fR \fICIPHER-SUITE-ID\fR, \fB\-\-cipher\-suite-id\fR=\fICIPHER-SUITE-ID\fR
Specify the IPMI 2.0 cipher suite ID to use. The Cipher Suite ID
identifies a set of authentication, integrity, and confidentiality
algorithms to use for IPMI 2.0 communication. The authentication
algorithm identifies the algorithm to use for session setup, the
integrity algorithm identifies the algorithm to use for session packet
signatures, and the confidentiality algorithm identifies the algorithm
to use for payload encryption. Defaults to cipher suite ID 3 if not
specified. The following cipher suite ids are currently supported:
.sp
0 - Authentication Algorithm = None; Integrity Algorithm = None; Confidentiality Algorithm = None
.sp
1 - Authentication Algorithm = HMAC-SHA1; Integrity Algorithm = None; Confidentiality Algorithm = None
.sp
2 - Authentication Algorithm = HMAC-SHA1; Integrity Algorithm = HMAC-SHA1-96; Confidentiality Algorithm = None
.sp
3 - Authentication Algorithm = HMAC-SHA1; Integrity Algorithm = HMAC-SHA1-96; Confidentiality Algorithm = AES-CBC-128
.\" .sp
.\" 4 - Authentication Algorithm = HMAC-SHA1; Integrity Algorithm = HMAC-SHA1-96; Confidentiality Algorithm = xRC4-128
.\" .sp
.\" 5 - Authentication Algorithm = HMAC-SHA1; Integrity Algorithm = HMAC-SHA1-96; Confidentiality Algorithm = xRC4-40
.sp
6 - Authentication Algorithm = HMAC-MD5; Integrity Algorithm = None; Confidentiality Algorithm = None
.sp
7 - Authentication Algorithm = HMAC-MD5; Integrity Algorithm = HMAC-MD5-128; Confidentiality Algorithm = None
.sp
8 - Authentication Algorithm = HMAC-MD5; Integrity Algorithm = HMAC-MD5-128; Confidentiality Algorithm = AES-CBC-128
.\" .sp
.\" 9 - Authentication Algorithm = HMAC-MD5; Integrity Algorithm = HMAC-MD5-128; Confidentiality Algorithm = xRC4-128
.\" .sp
.\" 10 - Authentication Algorithm = HMAC-MD5; Integrity Algorithm = HMAC-MD5-128; Confidentiality Algorithm = xRC4-40
.sp
11 - Authentication Algorithm = HMAC-MD5; Integrity Algorithm = MD5-128; Confidentiality Algorithm = None
.sp
12 - Authentication Algorithm = HMAC-MD5; Integrity Algorithm = MD5-128; Confidentiality Algorithm = AES-CBC-128
.\" .sp
.\" 13 - Authentication Algorithm = HMAC-MD5; Integrity Algorithm = MD5-128; Confidentiality Algorithm = xRC4-128
.\" .sp
.\" 14 - Authentication Algorithm = HMAC-MD5; Integrity Algorithm = MD5-128; Confidentiality Algorithm = xRC4-40
.\" XXX GUESS
.\" .sp
.\" 15 - Authentication Algorithm = HMAC-SHA256; Integrity Algorithm = None; Confidentiality Algorithm = None
.\" XXX GUESS
.\" .sp
.\" 16 - Authentication Algorithm = HMAC-SHA256; Integrity Algorithm = HMAC_SHA256_128; Confidentiality Algorithm = None
.sp
17 - Authentication Algorithm = HMAC-SHA256; Integrity Algorithm = HMAC_SHA256_128; Confidentiality Algorithm = AES-CBC-128
.\" XXX GUESS
.\" .sp
.\" 18 - Authentication Algorithm = HMAC-SHA256; Integrity Algorithm = HMAC_SHA256_128; Confidentiality Algorithm = xRC4-128
.\" XXX GUESS
.\" .sp
.\" 19 - Authentication Algorithm = HMAC-SHA256; Integrity Algorithm = HMAC_SHA256_128; Confidentiality Algorithm = xRC4-40
.TP
\fB\-l\fR \fIPRIVILEGE\-LEVEL\fR, \fB\-\-privilege\-level\fR=\fIPRIVILEGE\-LEVEL\fR
Specify the privilege level to be used. The currently available
privilege levels are USER, OPERATOR, and ADMIN. Defaults to OPERATOR
if not specified.
.TP
\fB\-\-config\-file\fR=\fIFILE\fR
Specify an alternate configuration file.
.TP
\fB\-W\fR \fIWORKAROUNDS\fR, \fB\-\-workaround\-flags\fR=\fIWORKAROUNDS\fR
Specify workarounds to vendor compliance issues. Multiple workarounds
can be specified separated by commas. A special command line flag of
"none", will indicate no workarounds (may be useful for overriding
configured defaults). See WORKAROUNDS below for a list of available
workarounds.
.TP
\fB\-\-debug\fR
Turn on debugging.
.TP
\fB\-?\fR, \fB\-\-help\fR
Output a help list and exit.
.TP
\fB\-\-usage\fR
Output a usage message and exit.
.TP
\fB\-V\fR, \fB\-\-version\fR
Output the program version and exit.
.if 0 \{
.TP
\fB\-\-rmcpdump\fR
Turn on RMCP packet dump output. Warning, the dump output can get
extremely long.
\}
.SH "IPMIPOWER OPTIONS"
The following options are specific to
.B ipmipower.
.TP
\fB\-n\fR, \fB\-\-on\fR
Power on the target hosts.
.TP
\fB\-f\fR, \fB\-\-off\fR
Power off the target hosts.
.TP
\fB\-c\fR, \fB\-\-cycle\fR
Power cycle the target hosts.
.TP
\fB\-r\fR, \fB\-\-reset\fR
Reset the target hosts.
.TP
\fB\-s\fR, \fB\-\-stat\fR
Get power status of the target hosts.
.TP
\fB\-\-pulse\fR
Send power diagnostic interrupt to target hosts.
.TP
\fB\-\-soft\fR
Initiate a soft-shutdown of the OS via ACPI.
.TP
\fB\-\-on\-if\-off\fR
The IPMI specification does not require the power cycle or hard reset
commands to turn on a machine that is currently powered off. This
option will force
.B ipmipower
to issue a power on command instead of a power cycle or hard reset
command if the remote machine's power is currently off.
.TP
\fB\-\-wait\-until\-on\fR
The IPMI specification allows power on commands to return prior to the
power on actually taking place. This option will force
.B ipmipower
to regularly query the remote BMC and return only after the machine
has powered on.
.TP
\fB\-\-wait\-until\-off\fR
The IPMI specification allows power off commands to return prior the
power off actually taking place. This option will force
.B ipmipower
to regularly query the remote BMC and return only after the machine
has powered off.
.SH "IPMIPOWER ADVANCED NETWORK OPTIONS"
The following options are used to change the networking behavior of
.B ipmipower.
.TP
\fB\-\-retransmission\-wait\-timeout\fR=\fIMILLISECONDS\fR
Specify the retransmission wait timeout length in milliseconds. The
retransmission wait timeout is similar to the retransmission timeout
above, but is used specifically for power completion verification with
the \fB\-\-wait\-until\-on\fR and \fB\-\-wait\-until\-off\fR options.
Defaults to 500 milliseconds (0.5 seconds).
.TP
\fB\-\-retransmission\-backoff\-count\fR=\fICOUNT\fR
Specify the retransmission backoff count for retransmissions. After
ever COUNT retransmissions, the retransmission timeout length will be
increased by another factor. Defaults to 8.
.TP
\fB\-\-ping\-interval\fR=\fIMILLISECONDS\fR
Specify the ping interval length in milliseconds. When running in
interactive mode, RMCP (Remote Management Control Protocol) discovery
messages will be sent to all configured remote hosts every
MILLISECONDS to confirm their support of IPMI. Power commands cannot
be sent to a host until it is discovered (or re-discovered if
previously lost). Defaults to 5000 milliseconds (5 seconds). Ping
discovery messages can be disabled by setting this valu to 0. RMCP
ping discovery messages are automatically disabled in non-interactive
mode.
.TP
\fB\-\-ping\-timeout\fR=\fIMILLISECONDS\fR
Specify the ping timeout length in milliseconds. When running in
interactive mode, RMCP (Remote Management Control Protocol) messages
discovery will be sent to all configured remote hosts to confirm their
support of IPMI. A remote host is considered undiscovered if the host
does not respond in MILLISECONDS time. Defaults to 30000 milliseconds
(30 seconds). The ping timeout cannot be larger than the ping interval.
.TP
\fB\-\-ping\-packet\-count\fR=\fICOUNT\fR
Specify the ping packet count size. Defaults to 10. See the
\fB\-\-ping-percent\-fR option below for more information on this
option.
.TP
\fB\-\-ping\-percent\fR=\fIPERCENT\fR
Specify the ping percent value. Defaults to 50.
Since IPMI is based on UDP, it is
difficult for
.B ipmipower
to distinguish between a missing machine and a bad (or heavily loaded)
network connection in interactive mode. when running in interactive
mode. For example, suppose a link consistently drops 80% of the
packets to a particular machine. The power control operation may have
difficulty completing, although a recent pong response from RMCP makes
.B ipmipower
believe the machine is up and functioning properly.
The ping packet acount and percent options are used to alleviate this
problem.
.B Ipmipower
will monitor RMCP ping packets in packet count chunks. If
.B ipmipower
does not receive a response to greater than ping percent of those
packets,
.B ipmipower
will assume the link to this node is bad and will not send power
control operations to that node until the connection is determined to
be reliable. This heuristic can be disabled by setting either the
ping packet count or ping percent to 0. This feature is not used if
ping interval is set to 0.
.TP
\fB\-\-ping\-consec\-count\fR=\fICOUNT\fR
Specify the ping consecutive count. This is another heuristic used to
determine if a node should be considered discovered, undiscovered, or
with a bad connection. If a valid RMCP pong response was received for
the last COUNT ping packets, a node will be considered discovered,
regardless of other heuristics listed above. Defaults to 5. This
heuristic can be disabled by setting this value to 0. This feature is
not used if other ping features described above are disabled.
.LP
.SH "HOSTRANGED OPTIONS"
The following options manipulate hostranged output. See HOSTRANGED
SUPPORT below for additional information on hostranges.
.TP
\fB\-B\fR, \fB\-\-buffer-output\fR
Buffer hostranged output. For each node, buffer standard output until
the node has completed its IPMI operation. When specifying this
option, data may appear to output slower to the user since the the
entire IPMI operation must complete before any data can be output.
See HOSTRANGED SUPPORT below for additional information.
.TP
\fB\-C\fR, \fB\-\-consolidate-output\fR
Consolidate hostranged output. The complete standard output from
every node specified will be consolidated so that nodes with identical
output are not output twice. A header will list those nodes with the
consolidated output. When this option is specified, no output can be
seen until the IPMI operations to all nodes has completed. If the
user breaks out of the program early, all currently consolidated
output will be dumped. See HOSTRANGED SUPPORT below for additional
information.
.TP
\fB\-F\fR \fINUM\fR, \fB\-\-fanout\fR=\fINUM\fR
Specify multiple host fanout. Indicates the maximum number of power
control operations that can be executed in parallel.
.TP
\fB\-E\fR, \fB\-\-eliminate\fR
Eliminate hosts determined as undetected by
.B ipmidetect.
This attempts to remove the common issue of hostranged execution
timing out due to several nodes being removed from service in a large
cluster. The
.B ipmidetectd
daemon must be running on the node executing the command.
.TP
\fB\-\-always\-prefix\fR

debian/NEWS
freeipmi (1.1.5-1) unstable; urgency=low

  * Config files were moved under /etc/freeipmi.  Legacy location (under
    /etc) is supported only for backwards compatibility.  Please migrate
    your changes and remove configuration files from legacy locations.
  * ipmimonitoring tool was removed -- superseded by ipmi-sensors
    and the new --output-sensor-state option (thin wrapper
    /usr/sbin/ipmimonitoring is provided)

 -- Yaroslav Halchenko <debian@onerussian.com>  Mon, 04 Jun 2012 15:46:08 -0400

freeipmi (0.8.10-1) experimental; urgency=low

  Upstream has switched from CVS to SVN.  Since target SVN seemed to lack any
  history of merging, I had decided to revert to classical way to package for
  Debian based on tarballs.  Thus do not expect complete development history
  upon checkout.  As a consequence I  have abandoned any previous history of
  Debian packaging, and recreated Debian packaging repository use
  git-import-dscs --debsnap.  Please remove any previous debcheckout of
  freeipmi you might have.

  To stay consistent with upstream, following init.d/default/logrotate.d
  files for freeipmi-bmc-watchdog were renamed accordingly (with handling
  of renames done by dpkg-maintscript-helper mv_conffile in debhelper
  scripts).
  Please verify that your changes to corresponding files were preserved.

 -- Yaroslav Halchenko <debian@onerussian.com>  Tue, 02 Nov 2010 00:42:46 -0400


ipmipower/src/ipmipower.c
/*****************************************************************************\
 *  $Id: ipmipower.c,v 1.72 2008/08/12 18:14:40 chu11 Exp $
 *****************************************************************************
 *  Copyright (C) 2007-2008 Lawrence Livermore National Security, LLC.
 *  Copyright (C) 2003-2007 The Regents of the University of California.
 *  Produced at Lawrence Livermore National Laboratory (cf, DISCLAIMER).
 *  Written by Albert Chu <chu11@llnl.gov>
 *  UCRL-CODE-155698
 *  
 *  This file is part of Ipmipower, a remote power control utility.
 *  For details, see http://www.llnl.gov/linux/.
 *  
 *  Ipmipower is free software; you can redistribute it and/or modify 
 *  it under the terms of the GNU General Public License as published by the 
 *  Free Software Foundation; either version 2 of the License, or (at your 
 *  option) any later version.
 *  
 *  Ipmipower is distributed in the hope that it will be useful, but 
 *  WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY 
 *  or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License 
 *  for more details.
 *  
 *  You should have received a copy of the GNU General Public License along
 *  with Ipmipower.  If not, see <http://www.gnu.org/licenses/>.
\*****************************************************************************/

#if HAVE_CONFIG_H
#include "config.h"
#endif /* HAVE_CONFIG_H */

#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#if HAVE_UNISTD_H
#include <unistd.h>
#endif /* HAVE_UNISTD_H */
#if STDC_HEADERS
#include <string.h>
#endif /* STDC_HEADERS */
#if TIME_WITH_SYS_TIME
#include <sys/time.h>
#include <time.h>
#else  /* !TIME_WITH_SYS_TIME */
#if HAVE_SYS_TIME_H
#include <sys/time.h>
#else /* !HAVE_SYS_TIME_H */
#include <time.h>
#endif  /* !HAVE_SYS_TIME_H */
#endif /* !TIME_WITH_SYS_TIME */
#include <stdint.h>
#include <sys/stat.h>
#include <sys/resource.h>
#if HAVE_FCNTL_H
#include <fcntl.h>
#endif /* HAVE_FCNTL_H */
#include <errno.h>

#include "ipmipower.h"
#include "ipmipower_argp.h"
#include "ipmipower_connection.h"
#include "ipmipower_powercmd.h"
#include "ipmipower_prompt.h"
#include "ipmipower_ping.h"
#include "ipmipower_util.h"
#include "ipmipower_wrappers.h"

#include "freeipmi-portability.h"
#include "tool-common.h"

cbuf_t ttyin;
cbuf_t ttyout;
cbuf_t ttyerr;

/* configuration for ipmipower */
struct ipmipower_arguments cmd_args;

/* Array of all ipmi "connections" */
struct ipmipower_connection *ics = NULL;
unsigned int ics_len = 0;

/* Array of hostlists for short output */
int output_hostrange_flag = 0;
hostlist_t output_hostrange[MSG_TYPE_NUM_ENTRIES];

/* _setup
 * - Setup structures and values for the program
 */
static void 
_setup(void) 
{
  int i;
  struct rlimit rlim;
  
  srand(Time(NULL));

  /* Make best effort to increase file descriptor limit, if it fails
   * for any reason, don't worry about it, its no big deal.
   */
  
  if (getrlimit(RLIMIT_NOFILE, &rlim) == 0)
    {
      rlim.rlim_cur = rlim.rlim_max;
      setrlimit(RLIMIT_NOFILE,&rlim);
    }

  /* Create TTY bufs */
  ttyin  = Cbuf_create(IPMIPOWER_MIN_TTY_BUF, IPMIPOWER_MAX_TTY_BUF);
  ttyout = Cbuf_create(IPMIPOWER_MIN_TTY_BUF, IPMIPOWER_MAX_TTY_BUF);
  ttyerr = Cbuf_create(IPMIPOWER_MIN_TTY_BUF, IPMIPOWER_MAX_TTY_BUF);

  for (i = 0; i < MSG_TYPE_NUM_ENTRIES; i++) 
    {
      if (!(output_hostrange[i] = hostlist_create(NULL)))
        ierr_exit("hostlist_create() error");
    }
  
  /* in interactive mode errors should always go to atleast the syslog */
  ierr_syslog(1);

#ifndef NDEBUG
  /* if debug set, send debug info to stderr too */
  ierr_cbuf(cmd_args.common.debug, ttyerr);

  /* on ierr_exit() dump cbuf data to appropriate places too */
  ierr_cbuf_dump_file_stream(cmd_args.common.debug, stderr);
#else  /* !NDEBUG */
  ierr_cbuf(0, 0);
#endif /* !NDEBUG */
}

/* _cleanup
 * - Cleanup dynamic memory allocated in this program
 */
static void 
_cleanup(void) 
{
  int i;

  cbuf_destroy(ttyin);
  
  /* Flush before destroying. */
  cbuf_read_to_fd(ttyout, STDOUT_FILENO, -1);
  cbuf_destroy(ttyout);
  cbuf_read_to_fd(ttyerr, STDERR_FILENO, -1);
  cbuf_destroy(ttyerr);

  ipmipower_connection_array_destroy(ics, ics_len);

  for (i = 0; i < MSG_TYPE_NUM_ENTRIES; i++)
    hostlist_destroy(output_hostrange[i]);
}

static void 
_sendto(cbuf_t buf, int fd, struct sockaddr_in *destaddr) 
{
  int n, rv;
  char buffer[IPMIPOWER_PACKET_BUFLEN];

  if ((n = cbuf_read(buf, buffer, IPMIPOWER_PACKET_BUFLEN)) < 0)
    ierr_exit("_sendto(%d): cbuf_read: %s", fd, strerror(errno));
  if (n == IPMIPOWER_PACKET_BUFLEN)
    ierr_exit("_sendto: Buffer full");
    
  rv = ipmi_lan_sendto(fd, buffer, n, 0, (struct sockaddr *)destaddr,
                       sizeof(struct sockaddr_in));
  if (rv < 0)
    ierr_exit("_sendto: ipmi_lan_sendto %s", strerror(errno));
  if (rv != n)
    ierr_exit("_sendto: ipmi_lan_sendto rv=%d n=%d", rv, n);
  
  /* cbuf should be empty now */
  if (!cbuf_is_empty(buf))
    ierr_exit("_sendto: cbuf not empty");
}

static void 
_recvfrom(cbuf_t buf, int fd, struct sockaddr_in *srcaddr) 
{
  int n, rv, dropped = 0;
  char buffer[IPMIPOWER_PACKET_BUFLEN];
  struct sockaddr_in from;
  unsigned int fromlen = sizeof(struct sockaddr_in);
  
  rv = ipmi_lan_recvfrom(fd, buffer, IPMIPOWER_PACKET_BUFLEN, 0, 
			 (struct sockaddr *)&from, &fromlen);
  if (rv < 0)
    ierr_exit("_recvfrom: ipmi_lan_recvfrom: %s", strerror(errno));
  if (!rv)
    ierr_exit("_recvfrom: ipmi_lan_recvfrom: EOF");
  
  /* Don't store if this is packet is strange for some reason */
  if (from.sin_family != AF_INET 
      || from.sin_addr.s_addr != srcaddr->sin_addr.s_addr)
    return;
  
  /* cbuf should be empty, but if it isn't, empty it */
  if (!cbuf_is_empty(buf)) 
    {
      ierr_dbg("_recvfrom: cbuf not empty, draining");
      do 
        {
          char tempbuf[IPMIPOWER_PACKET_BUFLEN];
          if (cbuf_read(buf, tempbuf, IPMIPOWER_PACKET_BUFLEN) < 0)
            ierr_exit("_recvfrom: cbuf_read: %s", strerror(errno));
        } while(!cbuf_is_empty(buf));
    }

  if ((n = cbuf_write(buf, buffer, rv, &dropped)) < 0)
    ierr_exit("_recvfrom(%d): cbuf_write: %s", fd, strerror(errno));
  if (n != rv)
    ierr_exit("_recvfrom: rv=%d n=%d", rv, n);
  if (dropped)
    ierr_dbg("_recvfrom: read dropped %d bytes", dropped);
}

/* _poll_loop
 * - poll on all descriptors
 */
static void 
_poll_loop(int non_interactive) 
{
  int nfds = 0;
  struct pollfd *pfds = NULL;

  while (non_interactive || ipmipower_prompt_process_cmdline()) 
    {
      int i, num, timeout;
      int powercmd_timeout = -1;
      int ping_timeout = -1;
      
      /* If there are no pending commands before this call,
       * powercmd_timeout will not be set, leaving it at -1
       */
      num = ipmipower_powercmd_process_pending(&powercmd_timeout);
      if (non_interactive && !num)
        break;

      /* ping timeout is always set if cmd_args.ping_interval > 0 */
      ipmipower_ping_process_pings(&ping_timeout);

      if (cmd_args.ping_interval) 
        {
          if (powercmd_timeout == -1)
            timeout = ping_timeout;
          else 
            timeout = (ping_timeout < powercmd_timeout) ? 
              ping_timeout : powercmd_timeout;
        }
      else
        timeout = powercmd_timeout; 
      
      /* achu: I always wonder if this poll() loop could be done far
       * more elegantly and efficiently without all this crazy
       * indexing, perhaps through a callback/event mechanism.  It'd
       * probably be more efficient, since most callback/event based
       * models have min-heap like structures inside for determining
       * what things timed out. Overall though, I don't think the O(n)
       * (n being hosts/fds) processing is really that inefficient for
       * this particular application and is not worth going back and
       * changing.  By going to a callback/event mechanism, there will
       * still be some O(n) activities within the code, so I am only
       * going to create a more efficient O(n) poll loop.
       */

      /* Has the number of hosts changed? */
      if (nfds != (ics_len*2) + 3)
	{
	  /* The "*2" is for each host's two fds, one for ipmi
	   * (ipmi_fd) and one for rmcp (ping_fd).  The "+3" is for
	   * stdin, stdout, stderr.
	   */
	  nfds = (ics_len*2) + 3;   
	  Free(pfds);
	  pfds = (struct pollfd *)Malloc(nfds * sizeof(struct pollfd));
	}
      
      for (i = 0; i < ics_len; i++) 
        {
          pfds[i*2].fd = ics[i].ipmi_fd;
          pfds[i*2+1].fd = ics[i].ping_fd;
          pfds[i*2].events = pfds[i*2+1].events = 0;
          pfds[i*2].revents = pfds[i*2+1].revents = 0;
          
          pfds[i*2].events |= POLLIN;
          if (!cbuf_is_empty(ics[i].ipmi_out))
            pfds[i*2].events |= POLLOUT;
          
          if (!cmd_args.ping_interval)
            continue;
          
          pfds[i*2+1].events |= POLLIN;
          if (!cbuf_is_empty(ics[i].ping_out))
            pfds[i*2+1].events |= POLLOUT;
        }
      
      pfds[nfds-3].fd = STDIN_FILENO;
      pfds[nfds-3].events = POLLIN;
      pfds[nfds-3].revents = 0;
      pfds[nfds-2].fd = STDOUT_FILENO;
      if (!cbuf_is_empty(ttyout))
        pfds[nfds-2].events = POLLOUT;
      else
        pfds[nfds-2].events = 0;
      pfds[nfds-2].revents = 0;
      pfds[nfds-1].fd = STDERR_FILENO;
      if (!cbuf_is_empty(ttyerr))
        pfds[nfds-1].events = POLLOUT;
      else
        pfds[nfds-1].events = 0;
      pfds[nfds-1].revents = 0;

      Poll(pfds, nfds, timeout);
      
      for (i = 0; i < ics_len; i++) 
        {
          if (pfds[i*2].revents & POLLERR) 
            {
              ierr_dbg("_poll_loop: IPMI POLLERR, %s, %d", ics[i].hostname, 
                       ics[i].ipmi_fd);
              continue;
            }
          if (pfds[i*2].revents & POLLIN) 
            _recvfrom(ics[i].ipmi_in, ics[i].ipmi_fd, &(ics[i].destaddr));
          if (pfds[i*2].revents & POLLOUT)
            _sendto(ics[i].ipmi_out, ics[i].ipmi_fd, &(ics[i].destaddr));
          
          if (!cmd_args.ping_interval)
            continue;
          
          if (pfds[i*2+1].revents & POLLERR) 
            {
              ierr_dbg("_poll_loop: PING POLLERR, %s, %d", ics[i].hostname, 
                       ics[i].ipmi_fd);
              continue;
            }
          if (pfds[i*2+1].revents & POLLIN) 
            _recvfrom(ics[i].ping_in, ics[i].ping_fd, &(ics[i].destaddr));
          if (pfds[i*2+1].revents & POLLOUT)
            _sendto(ics[i].ping_out, ics[i].ping_fd, &(ics[i].destaddr));
        }
      
      if (pfds[nfds-3].revents & POLLIN)
        Cbuf_write_from_fd(ttyin, STDIN_FILENO);
      if (!cbuf_is_empty(ttyout) && (pfds[nfds-2].revents & POLLOUT))
        Cbuf_read_to_fd(ttyout, STDOUT_FILENO);
      if (!cbuf_is_empty(ttyerr) && (pfds[nfds-1].revents & POLLOUT))
        Cbuf_read_to_fd(ttyerr, STDERR_FILENO);
    }

  Free(pfds);
}

static void
_eliminate_nodes(void)
{
  if (cmd_args.hostrange.eliminate)
    {
      ipmidetect_t id = NULL;
      int i;

      if (!(id = ipmidetect_handle_create()))
        ierr_exit("ipmidetect_handle_create");
      
      if (ipmidetect_load_data(id,
                               NULL,
                               0,
                               0) < 0)
        {
          if (ipmidetect_errnum(id) == IPMIDETECT_ERR_CONNECT
              || ipmidetect_errnum(id) == IPMIDETECT_ERR_CONNECT_TIMEOUT)
            ierr_exit("Error connecting to ipmidetect daemon");
          ierr_exit("ipmidetect_load_data: %s", ipmidetect_errormsg(id));
        }
      
      for (i = 0; i < ics_len; i++)
        {
          int ret;
          
          if ((ret = ipmidetect_is_node_detected(id, ics[i].hostname)) < 0)
            {
              if (ipmidetect_errnum(id) == IPMIDETECT_ERR_NOTFOUND)
                ierr_exit("Node '%s' unrecognized by ipmidetect", ics[i].hostname);
              ierr_exit("ipmidetect_is_node_detected: %s", ipmidetect_errormsg(id));
            }

          if (!ret)
            ics[i].skip++;
        }
      
      ipmidetect_handle_destroy(id);
    }
}

int 
main(int argc, char *argv[]) 
{
  /* Call before anything else */
  ierr_init(argv[0]);
  ierr_file_descriptor(1, STDERR_FILENO); /* initially errors goto stderr */
  
  ipmi_disable_coredump();

  if (ipmi_rmcpplus_init() < 0)
    ierr_exit("ipmi_rmcpplus_init");

  ipmipower_argp_parse (argc, argv, &cmd_args);

  _setup();

  ipmipower_powercmd_setup();
  
  if (cmd_args.common.hostname) 
    {
      unsigned int len = 0;
      if (!(ics = ipmipower_connection_array_create(cmd_args.common.hostname, &len)))
        {
          /* dump error outputs here, most notably invalid hostname output */
          if (cbuf_read_to_fd(ttyout, STDOUT_FILENO, -1) > 0)
            exit(1);
          ierr_exit("ipmipower_connection_array_create: %s", strerror(errno));
        }
      ics_len = len;
    }
  
  /* If power command (i.e. --reset, --stat, etc.) is passed at
   * command line, put the power control commands in the pending
   * queue.
   */
  if (cmd_args.powercmd != POWER_CMD_NONE) 
    {
      int i;

      cmd_args.ping_interval = 0;

      /* Check for appropriate privilege first */
      if (cmd_args.common.privilege_level == IPMI_PRIVILEGE_LEVEL_USER 
          && POWER_CMD_REQUIRES_OPERATOR_PRIVILEGE_LEVEL(cmd_args.powercmd))
        ierr_exit("power operation requires atleast operator privilege");

      _eliminate_nodes();

      for (i = 0; i < ics_len; i++) 
        {
          if (ics[i].skip)
            continue;
          ipmipower_powercmd_queue(cmd_args.powercmd, &ics[i]);
        }
    }
  
  ierr_file_descriptor(0, 0); /* now errors are done through the ttyerr */

  /* immediately send out discovery messages upon startup */
  ipmipower_ping_force_discovery_sweep();

  _poll_loop((cmd_args.powercmd != POWER_CMD_NONE) ? 1 : 0);
  
  ipmipower_powercmd_cleanup();
  _cleanup();
  exit(0);
}
